### PR TITLE
Create TransactionRepository class

### LINKED ISSUE #27: Iteration 3.3 Transaction Repository
TransactionRepository
Transactions are billing records for an invoice. An invoice can have multiple transactions, but should have at most one that is successful.

The TransactionRepository is responsible for holding and searching our Transaction instances. It offers the following methods:

all - returns an array of all known Transaction instances
find_by_id - returns either nil or an instance of Transaction with a matching ID
find_all_by_invoice_id - returns either [] or one or more matches which have a matching invoice ID
find_all_by_credit_card_number - returns either [] or one or more matches which have a matching credit card number
find_all_by_result - returns either [] or one or more matches which have a matching status
create(attributes) - create a new Transaction instance with the provided attributes. The new Transaction’s id should be the current highest Transaction id plus 1.
update(id, attribute) - update the Transaction instance with the corresponding id with the provided attributes. Only the transaction’s credit_card_number, credit_card_expiration_date, and result can be updated. This method will also change the transaction’s updated_at attribute to the current time.
delete(id) - delete the Transaction instance with the corresponding id
The data can be found in data/transactions.csv so the instance is created and used like this:

```rb
sales_engine = SalesEngine.from_csv(:transactions => "./data/transactions.csv")
transaction = sales_engine.transactions.find_by_id(6)
# => <Transaction>
``` 

3 points

### PR DESCRIPTION
**Pull request by mkrumholz**
### Changes

- Description of changes
Created TransactionRepository class with test file. Uses custom mocks for all tests. 

### Requested review on...
- Specifics on, logic, naming schemes, etc.

### Checklist
- [x] Tests pass
- [ ] Ready to merge

closes #27

### CODE REVIEW

File: lib/transaction.rb [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+              :result,
+              :created_at,
+              :updated_at
+
+  def initialize(details)
+    @id = details[:id]
+    @invoice_id = details[:invoice_id]
+    @credit_card_number = details[:credit_card_number]
+    @credit_card_expiration_date = details[:credit_card_expiration_date]
+    @result = details[:result].to_sym
```
Comment (NoahZinter) [inline lib/transaction.rb:15]: Nice catch here, I only saw @result needing to be a symbol in the spec harness

**Comment by mkrumholz** [comment]
Ugh wait a bit on reviewing here. Need to add guards to my update statements for the spec harness.

File: spec/transaction_repository_spec.rb [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+        created_at: Time.new(2021, 0o2, 0o4),
+        updated_at: Time.new(2021, 0o2, 0o4)
+      }
+      t_repo.update(7, new_info)
+      updated_item = t_repo.transactions.last
+
+      expect(updated_item.id).to eq 7
+      expect(updated_item.invoice_id).to eq 8
+      expect(updated_item.created_at).to eq Time.new(2021, 0o3, 0o1)
+      expect(updated_item.updated_at).to be > Time.new(2021, 0o3, 0o1)
```
Comment (NoahZinter) [inline spec/transaction_repository_spec.rb:201]: I like this Time test, I'm about to lift it from the spec harness for my repo class

**Review by NoahZinter — APPROVED** [review]
Looks great!

**Review by redferret — COMMENTED** [review]
Nice workaround with the MockData.
Do you think it's necessary to incorporate those mock methods or does it feel clean to just leave it using the `generate_mock` method?

**Comment by mkrumholz** [comment]
> Nice workaround with the MockData.
> Do you think it's necessary to incorporate those mock methods or does it feel clean to just leave it using the `generate_mock` method?

Honestly, I don't think it looks half bad this way and it all works just fine. I'd leave it as-is for now and we can come back to it if there's ample time at the end

**Review by redferret — APPROVED** [review]
Looks good, could you remove those comments about the frozen strings?

**Comment by mkrumholz** [comment]
> Looks good, could you remove those comments about the frozen strings?

Done! Should be good to go now.

**Pull request merged**